File: 015.3Sum/wyrj/3Sum.c
/* 24ms */
/**
 * Return an array of arrays of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int cmp(const void *a, const void *b) {
    int *aa = (int *)a;
    int *bb = (int *)b;
    if (*aa > *bb) {
        return 1;
    } else if (*aa < *bb) {
        return -1;
    } else {
        return 0;
    }
}
int *createElement(int a, int b, int c) {
    int *el = (int *)malloc(sizeof(int) * 3);
    el[0] = a;
    el[1] = b;
    el[2] = c;
    return el;
}
int** threeSum(int* nums, int numsSize, int* returnSize) {
    int **ret = (int **)malloc(sizeof(int *)*numsSize*numsSize);
    qsort(nums, numsSize, sizeof(int), cmp);
    int size = 0;
    int i, j, k;
    for (i = 0; i < numsSize - 2; i++) {
        if (0 < nums[i]) {
            break;
        }
        if (i > 0 && nums[i] == nums[i-1]) {
            continue;
        }
        k = numsSize - 1;
        for (j = i + 1; j < k; j++) {
            if (j != i + 1 && nums[j] == nums[j-1]) {
                continue;
            }
            while((0 < nums[i] + nums[j] + nums[k]) && k > j + 1) {
                k--;
            }
            if (0 == nums[i] + nums[j] + nums[k]) {
                ret[size++] = createElement(nums[i], nums[j], nums[k]);
            }
        }
    }
    *returnSize = size;
    return ret;
}

File: 015.3Sum/wyrj/3Sum.js
/* 280ms */
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    var cmp = function(a,b) {
        if (a > b) {
            return 1;
        } else if (a < b) {
            return -1;
        } else {
            return 0;
        }
    };
    nums.sort(cmp);
    var ret = [], index, i, j, nonPosIndex, nonNegIndex;
    for (i = 0; i < nums.length; i++) {
        if ((0 > nums[i-1] || 0 === i) && 0 <= nums[i]) {
            nonNegIndex = i;
        }
        if (0 >= nums[i] && (0 < nums[i+1] || nums.length - 1 === i)) {
            nonPosIndex = i;
        }
    }
    if (2 <= nonPosIndex - nonNegIndex) {
        ret.push([0,0,0]);
    }
    for (i = 0; i < nonNegIndex; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }
        index = nums.length - 1;
        for (j = i + 1; j < index; j++) {
            if (j !== i + 1 && nums[j] === nums[j-1]) {
                continue;
            }
            while((0 < nums[index] + nums[i] + nums[j]) && index > j) {
                index--;
            }
            if (j === index) {
                break;
            }
            if (0 === nums[index] + nums[i] + nums[j]) {
                ret.push([nums[i], nums[j], nums[index]]);
                index--;
            }
        }
    }
    return ret;
};